clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -212,25 +214,39 @@ public RecordAppendResult append(TopicPartition tp,
      * Abort the batches that have been sitting in RecordAccumulator for more than the configured requestTimeout
      * due to metadata being unavailable
      */
-    public List<RecordBatch> abortExpiredBatches(int requestTimeout, Cluster cluster, long now) {
+    public List<RecordBatch> abortExpiredBatches(int requestTimeout, Metadata metadata, long now) {
         List<RecordBatch> expiredBatches = new ArrayList<RecordBatch>();
         int count = 0;
         for (Map.Entry<TopicPartition, Deque<RecordBatch>> entry : this.batches.entrySet()) {
             Deque<RecordBatch> dq = entry.getValue();
-            synchronized (dq) {
-                // iterate over the batches and expire them if they have stayed in accumulator for more than requestTimeOut
-                Iterator<RecordBatch> batchIterator = dq.iterator();
-                while (batchIterator.hasNext()) {
-                    RecordBatch batch = batchIterator.next();
-                    // check if the batch is expired
-                    if (batch.maybeExpire(requestTimeout, now, this.lingerMs)) {
-                        expiredBatches.add(batch);
-                        count++;
-                        batchIterator.remove();
-                        deallocate(batch);
-                    } else {
-                        if (!batch.inRetry()) {
-                            break;
+            // Only expire a batch if its metadata is not available.
+            Cluster cluster = metadata.fetch();
+            // Check if the metadata might be old.
+            boolean maybeStaleMetadata = metadata.lastRefresh() > metadata.lastSuccessfulUpdate();
```

Comment:
@junrao We wanted this check because of https://issues.apache.org/jira/browse/KAFKA-2805.
In that scenario, if all the brokers went down, the old metadata will be in the producer and no batches will be timed out. In that case, metadata will be available, but leader is not connected. Since the accumulator has no idea about the leader connectivity, the only way it detects this situation is to check the metadata age.

The current patch tries to:
1. Timeout batches whose metadata is missing when metadata is not stale (i.e. no metadata update failure)
2. Timeout all the batches when metadata is stale enough (i.e. metadata update keeps failing for request timeout)